components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/GeckoLoginStorageDelegate.kt

```diff
@@ -0,0 +1,173 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.sync.logins
+
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import mozilla.components.concept.storage.Login
+import mozilla.components.concept.storage.LoginStorageDelegate
+import mozilla.components.lib.dataprotect.SecureAbove22Preferences
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * A type of persistence operation, either 'create' or 'update'.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+enum class Operation { CREATE, UPDATE }
+
+internal const val PASSWORDS_KEY = "passwords"
+
+/**
+ * [LoginStorageDelegate] implementation.
+ *
+ * An abstraction that handles the persistence and retrieval of [LoginEntry]s so that Gecko doesn't
+ * have to.
+ *
+ * In order to use this class, attach it to the active [GeckoRuntime] as its `loginStorageDelegate`.
+ * It is not designed to work with other engines.
+ *
+ * This class is part of a complex flow integrating Gecko and Application Services code, which is
+ * described here:
+ *
+ * - GV finds something on a page that it believes could be autofilled
+ * - GV calls [onLoginFetch]
+ *   - We retrieve [Login]s from all matching domains (if any) from [loginStorage]
+ *   - We return these [Login]s to GV
+ * - User submits their credentials
+ * - GV decides to save the login
+ *   - GV calls [onLoginSave] with a [Login]
+ *     - If this [Login] was autofilled, it is updated with new information but retains the same
+ *     [Login.guid]
+ *     - If this is a new [Login], its [Login.guid] will be an empty string
+ *   - We [CREATE] or [UPDATE] the [Login] in [loginStorage]
+ *     - If [CREATE], [loginStorage] generates a new guid for it
+ */
+class GeckoLoginStorageDelegate(
+    private val loginStorage: AsyncLoginsStorage,
+    keyStore: SecureAbove22Preferences,
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+) : LoginStorageDelegate {
+
+    private val password = { scope.async { keyStore.getString(PASSWORDS_KEY)!! } }
+    private val logger = Logger("GeckoLoginStorageDelegate")
+
+    override fun onLoginUsed(login: Login) {
+        val guid = login.guid
+        // If the guid is null, we have no way of associating the login with any record in the DB
+        if (guid == null || guid.isEmpty()) return
+        scope.launch {
+            loginStorage.withUnlocked(password) {
+                loginStorage.touch(guid).await()
+            }
+        }
+    }
+
+    override fun onLoginFetch(domain: String): List<Login> {
+        return runBlocking {
+            // GV expects a synchronous response. Blocking here hasn't caused problems during
+            // testing, but we should keep an eye on telemetry
```

Comment:
Which telemetry? How will you track if this is a perf bottleneck? Specifically, unlocking the store - not much you can easily do about storage access, other than pre-caching... E.g. this seems to be in the hot-path of toolkit code autofilling stuff on the page.

Reply:
Filed https://github.com/mozilla-mobile/android-components/issues/5531 to add a probe here.